File: 0566. Reshape the Matrix/0566.java
// time complexity: O(m*n)
// space complexity: O(m*n)
class Solution {
    public int[][] matrixReshape(int[][] mat, int r, int c) {
        // return original matrix if new dimensions are invalid
        if (r < 1 || c < 1) return mat;

        int row = mat.length;
        int column = mat[0].length;

        // return original matrix if total elements don't match
        if (r*c != row*column) return mat;

        // create new reshaped matrix
        int[][] reshaped = new int[r][c];

        int x = 0; // current row in original matrix
        int y = 0; // current column in original matrix

        // fill the reshaped matrix
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                reshaped[i][j] = mat[x][y];
                y++;
                // move to next row if end of column is reached
                if (y == column) {
                    y = 0;
                    x++;
                }
            }
        }

        return reshaped;
    }
}

File: 0566. Reshape the Matrix/0566.py
# time complexity: O(m*n)
# space complexity: O(m*n)
class Solution(object):
    def matrixReshape(self, mat, r, c):
        """
        :type mat: List[List[int]]
        :type r: int
        :type c: int
        :rtype: List[List[int]]
        """
        # dimensions of the original matrix
        original_rows = len(mat)
        original_cols = len(mat[0])

        # if the number of elements is not the same:
        # reshaping is not possible.
        if original_rows * original_cols != r * c:
            return mat

        # all elements to a 1D list
        elements = []
        for i in range(original_rows):
            for j in range(original_cols):
                elements.append(mat[i][j])

        # new matrix
        reshaped = []
        index = 0
        for i in range(r):
            new_row = []
            for j in range(c):
                new_row.append(elements[index])
                index += 1
            reshaped.append(new_row)

        return reshaped